drivers/usb/host/xhci.c

```diff
@@ -1385,24 +1385,59 @@ static int xhci_submit_control_msg(struct udevice *dev, struct usb_device *udev,
 		while (hop->parent->parent)
 			hop = hop->parent;
 */
-	return _xhci_submit_control_msg(udev, pipe, buffer, length, setup,
-					root_portnr);
+	if (sel4_dma_is_mapped(buffer))
```

Comment:
Presumably this has been found to be necessary; i.e. sometimes the buffer has already been mapped (and is freed elsewhere later) but sometimes the buffer has to be mapped, used, and freed all on-the-fly?
Applies equally to `_xhci_submit_control_msg`, `_xhci_submit_bulk_msg`, `_xhci_submit_int_msg`.

Reply:
Correct. This routine is used for buffers that are sometimes already within our DMA allocations (e.g. buffers declared within the XHCI stack itself) and sometimes are not DMA mapped (e.g. coming from the USB keyboard driver code).

When the buffer was already DMA mapped it is the responsibility of whichever piece of code that mapped it to tidy it up later. When we need to map it here ourselves then we have to do the house-keeping here.